### PR TITLE
AVRO-3345: Made Magic readonly in DataFileConstants

### PR DESCRIPTION
**Pull request by KyleSchoonover**
### Jira

- [ ] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3345

### Tests

- [ ] My PR does not need testing for this extremely good reason: non functional change

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/csharp/src/apache/main/File/DataFileConstants.cs [hunk 1]
```diff
@@ -64,10 +61,13 @@ public class DataFileConstants
         /// <summary>
         /// Magic bytes at the beginning of an Avro data file.
         /// </summary>
-        public static byte[] Magic = { (byte)'O',
-                                       (byte)'b',
-                                       (byte)'j',
-                                       Version };
+        public static readonly byte[] Magic =
```
Comment (martin-g) [inline lang/csharp/src/apache/main/File/DataFileConstants.cs:64]: The looks like an API change but I guess no one is supposed to change the magic bytes

File: lang/csharp/src/apache/main/File/DataFileConstants.cs [hunk 2]
```diff
@@ -21,9 +21,6 @@ namespace Avro.File
     /// <summary>
     /// Constants used in data files.
     /// </summary>
-    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design",
-        "CA1052:Static holder types should be Static or NotInheritable",
-        Justification = "Maintain public API")]
```
Comment (martin-g) [inline lang/csharp/src/apache/main/File/DataFileConstants.cs:26]: Is this change related to the `Magic` constant ?

**Review by martin-g — APPROVED** [review]

File: lang/csharp/src/apache/main/File/DataFileConstants.cs (hunk 2 above)
Comment (KyleSchoonover) [reply lang/csharp/src/apache/main/File/DataFileConstants.cs:26]: It was, but it's an old rule.  The way you get rid of it, is just make the static property readonly.  Allows you to essentially make a static property a constant.  The reason you do this is because constants are basic types and not a reference type like an array.

File: lang/csharp/src/apache/main/File/DataFileConstants.cs (hunk 1 above)
Comment (KyleSchoonover) [reply lang/csharp/src/apache/main/File/DataFileConstants.cs:64]: The change was formatting.

Comment (martin-g) [reply lang/csharp/src/apache/main/File/DataFileConstants.cs:64]: > The change was formatting.

What do you mean ?

AFAIU `readonly` makes `Magic` a constant, like `final` in Java.
Until now a user can change the value of `Magic` in his/her application. After this PR it won't be possible anymore. And such application will fail to compile. This is what I meant by API change.
But I agree that no one is supposed to change the magic bytes.

Comment (KyleSchoonover) [reply lang/csharp/src/apache/main/File/DataFileConstants.cs:64]: You are correct.  Do we want the backwards support to allow the change?  It appears the intention was that it is a constant hence the readonly static.  Like with most things, if it is available someone is using it.  The other option is I can mark this obsolete, and add the remark for a future change that it be made readonly.

Comment (martin-g) [reply lang/csharp/src/apache/main/File/DataFileConstants.cs:64]: I am 99% sure no one is setting custom value for the `Magic` field. So I am OK to merge it as is! But only for 1.12.0!
If someone complains that we broke his/her build then we can revert if (s)he gives us a really convincing reason!
